src/mito2/src/manifest/checkpointer.rs

```diff
@@ -0,0 +1,171 @@
+}
+
+impl Checkpointer {
+    pub(crate) fn new(
+        region_id: RegionId,
+        manifest_options: RegionManifestOptions,
+        manifest_store: ManifestObjectStore,
+        last_checkpoint_version: ManifestVersion,
+    ) -> Self {
+        Self {
+            manifest_options,
+            is_doing_checkpoint: Arc::new(AtomicBool::new(false)),
+            inner: Arc::new(Inner {
+                region_id,
+                manifest_store,
+                last_checkpoint_version: AtomicU64::new(last_checkpoint_version),
+            }),
+        }
+    }
+
+    pub(crate) fn last_checkpoint_version(&self) -> ManifestVersion {
+        self.inner.last_checkpoint_version.load(Ordering::Relaxed)
+    }
+
+    /// Check if it's needed to do checkpoint for the region by the checkpoint distance.
+    /// If needed, and there's no currently running checkpoint task, it will start a new checkpoint
+    /// task running in the background.
+    pub(crate) fn maybe_do_checkpoint(&self, manifest: &RegionManifest) {
+        if self.manifest_options.checkpoint_distance == 0 {
+            return;
+        }
+
+        let last_checkpoint_version = self.last_checkpoint_version();
+        if manifest.manifest_version - last_checkpoint_version
+            < self.manifest_options.checkpoint_distance
+        {
+            return;
+        }
+
+        // We can simply check whether there's a running checkpoint task like this, all because of
+        // the caller of this function is ran single threaded, inside the lock of RegionManifestManager.
+        if self.is_doing_checkpoint.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let start_version = if last_checkpoint_version == 0 {
+            // Checkpoint version can't be zero by implementation.
+            // So last checkpoint version is zero means no last checkpoint.
+            MIN_VERSION
+        } else {
+            last_checkpoint_version + 1
+        };
+        let end_version = manifest.manifest_version;
+        info!(
+            "Start doing checkpoint for region {}, compacted version: [{}, {}]",
+            self.inner.region_id(),
+            start_version,
+            end_version,
+        );
+
+        let checkpoint = RegionCheckpoint {
+            last_version: end_version,
+            compacted_actions: (end_version - start_version + 1) as usize,
+            checkpoint: Some(manifest.clone()),
+        };
+        self.do_checkpoint(checkpoint);
+    }
+
+    fn do_checkpoint(&self, checkpoint: RegionCheckpoint) {
+        let is_doing_checkpoint = self.is_doing_checkpoint.clone();
+        is_doing_checkpoint.store(true, Ordering::Relaxed);
+        let guard = scopeguard::guard(is_doing_checkpoint, |x| {
+            x.store(false, Ordering::Relaxed);
+        });
+
+        let inner = self.inner.clone();
+        common_runtime::spawn_bg(async move {
+            let _guard = guard;
+
+            inner.do_checkpoint(checkpoint).await;
```

Comment:
We move the scope guard into the async block eventually, so there is no need to put it in an individual `Arc`.
```suggestion
        let inner = scopeguard::guard(self.inner.clone(), |x| {
            x.is_doing_checkpoint.store(false, Ordering::Relaxed);
        });

        common_runtime::spawn_bg(async move {
            inner.is_doing_checkpoint.store(true, Ordering::Relaxed);
            inner.do_checkpoint(checkpoint).await;
```

Reply:
If setting the `is_doing_checkpoint` inside the async task, we can't be sure when it's been set. The next `maybe_do_checkpoint` could be called only to find it can do checkpoint again, which is not good.